Add support for the Web Extension's Event API in the UIProcess.

kiaraarose opened the pull request:
#### afed0b22500b9d8e73f1bc3bc28c381d2c95c956
<pre>
Add support for the Web Extension&apos;s Event API in the UIProcess.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248684.">https://bugs.webkit.org/show_bug.cgi?id=248684.</a>

Reviewed by Timothy Hatcher.

This patch adds support for the Event API in the UIProcess. This includes loading the background
page listeners from storage when the extension loads, and saving these listeners to storage
(for non-persistent background pages).

* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(-[_WKWebExtensionControllerConfiguration storageIsPersistent]):
Rename isPersistent() to storageIsPersistent() to make it clear that this function is for checking
to see if information relating to the extension is saved to storage and not for the extension
background page being persistent.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
(WebKit::WebExtensionContext::addListener):
(WebKit::WebExtensionContext::removeListener):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
Add FIXME to check to see if extension is being loaded as part of startup.

(WebKit::WebExtensionContext::stateFilePath const):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::writeStateToStorage const):
Use newly renamed storageIsPersistent().

(WebKit::WebExtensionContext::loadBackgroundWebViewDuringLoad):
Load background page listeners from storage.
Add FIXME to check to see if the extension is being loaded as part of startup.

(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
Retrieve background page listeners data from storage. Populate m_backgroundPageListeners.

(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
Save background page listeners to storage for non persistent background pages.

(WebKit::WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
Save background page listeners to storage.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::storageDirectory const):

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::storageIsPersistent const):
(WebKit::WebExtensionContext::isPersistent const): Deleted.

* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::storageIsPersistent const):
(WebKit::WebExtensionController::isPersistent const): Deleted.

* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp:
(WebKit::WebExtensionControllerConfiguration::copy const):
Use newly renamed storageIsPersistent().

* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::storageIsPersistent const):
(WebKit::WebExtensionControllerConfiguration::isPersistent const): Deleted.

Canonical link: <a href="https://commits.webkit.org/257589@main">https://commits.webkit.org/257589@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9890c2821a5d4e208ea523b36ed1936e8ca4a84

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99355 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32478 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108740 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9117 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85875 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91846 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106668 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105113 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90439 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33870 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21786 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2426 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23303 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2329 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [💥 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8490 "An unexpected error occured. Recent messages:Failed to print configuration") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42778 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2666 "Built successfully and passed tests") | [💥 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4197 "An unexpected error occured. Recent messages:Failed to print configuration") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -195,6 +200,9 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     // FIXME: <https://webkit.org/b/248430> Move local storage (if base URL changed).
 
+    // FIXME: <https://webkit.org/b/248889> Check to see if extension is being loaded as part of startup.
+    m_isBeingLoadAsPartOfBrowserLaunch = false;
```
Hmm, I'll have to think about this a little.

xeenon replied:
I think for now it is best to remove `m_isBeingLoadAsPartOfBrowserLaunch`. And just keep the FIXME.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1133,11 +1141,31 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
+    BOOL uniqueIdentifierHasChanged = ![m_uniqueIdentifier isEqualToString:lastSeenUniqueIdentifier()];
+    if (uniqueIdentifierHasChanged)
+        saveUniqueIdentifierToStorage();
```
The unique identifier in WebKit is not what it was in Safari. If the unique identifier changes in WebKit, all state will be reset, so there is no need to save this here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1133,11 +1141,31 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        BOOL backgroundPageListensToOnStartup = NO;
+        EventListenterTypeCountedSet savedBackgroundPageListeners = loadBackgroundPageListenersFromStorage();
+
+        if (savedBackgroundPageListeners.isEmpty() || backgroundPageListensToOnStartup || savedVersionNumberDoesNotMatchCurrentVersionNumber || uniqueIdentifierHasChanged || (backgroundPageListensToOnStartup && m_isBeingLoadAsPartOfBrowserLaunch))
```
`|| backgroundPageListensToOnStartup` should not be in here, it is checked at the end with `(backgroundPageListensToOnStartup && m_isBeingLoadAsPartOfBrowserLaunch)`.

xeenon replied:
Though, until `m_isBeingLoadAsPartOfBrowserLaunch` is valid, it likely should be just `|| backgroundPageListensToOnStartup`.

xeenon replied:
Add a FIXME here for <https://webkit.org/b/248889> Check to see if extension is being loaded as part of startup.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1133,11 +1141,31 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+            m_shouldFireStartupEvent = false;
+        }
+
+        m_isBeingLoadAsPartOfBrowserLaunch = false;
```
Remove for now.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
+WebExtensionContext::EventListenterTypeCountedSet WebExtensionContext::loadBackgroundPageListenersFromStorage()
+{
+    NSDictionary *extensionState = currentState();
```
I'd change this line to `NSData *listenersData = [currentState() objectForKey:backgroundPageListenersKey]` to shorten the next line.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     m_backgroundWebView = nil;
 }
 
+WebExtensionContext::EventListenterTypeCountedSet WebExtensionContext::loadBackgroundPageListenersFromStorage()
```
Instead of having this return a copy of the set, it would be better to just have it update `m_backgroundPageListeners`. You should call `m_backgroundPageListeners.clear()` at the start to make sure it does not duplicate data if it is called multiple times.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1133,11 +1141,31 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        if (savedBackgroundPageListeners.isEmpty() || backgroundPageListensToOnStartup || savedVersionNumberDoesNotMatchCurrentVersionNumber || uniqueIdentifierHasChanged || (backgroundPageListensToOnStartup && m_isBeingLoadAsPartOfBrowserLaunch))
+            loadBackgroundWebView();
+        else {
+            m_backgroundPageListeners = savedBackgroundPageListeners;
```
You can remove this once `loadBackgroundPageListenersFromStorage()` builds `m_backgroundPageListeners`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+NSString *WebExtensionContext::lastSeenUniqueIdentifier()
+{
+    return [m_state objectForKey:lastSeenUniqueIdentifierKey];
+}
```
Can be removed. See other comment.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    if (backgroundContentURL().isEmpty() || isPersistent())
+        return;
+
+    NSMutableDictionary *extensionState = [currentState() mutableCopy];
```
`m_state` is mutable, so you can just set keys on it.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    NSData *newBackgroundPageListenersAsData = [NSKeyedArchiver archivedDataWithRootObject:listeners requiringSecureCoding:YES error:nil];
+    NSData *savedBackgroundPageListenersAsData = [extensionState objectForKey:backgroundPageListenersKey];
+    extensionState[backgroundPageListenersKey] = newBackgroundPageListenersAsData;
```
Use `m_state.get()[backgroundPageListenersKey]`

kiaraarose replied:
I decided to do `[m_state objectForKey:backgroundPageListenersKey];`. Is there a preference of doing one over the other?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    extensionState[backgroundPageListenersKey] = newBackgroundPageListenersAsData;
+
+    NSNumber *savedListenerVersionNumber = [extensionState objectForKey:lastSeenVersionStateKey];
+    extensionState[lastSeenVersionStateKey] = @(currentBackgroundPageListenerStateVersion);
```
Ditto on using `m_state` directly.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    BOOL hasListenerStateChanged = ![newBackgroundPageListenersAsData isEqualToData:savedBackgroundPageListenersAsData];
+    BOOL hasVersionNumberChanged = savedListenerVersionNumber.integerValue != currentBackgroundPageListenerStateVersion;
+    if (hasListenerStateChanged || hasVersionNumberChanged) {
+        m_state = extensionState;
```
Not needed once you use `m_state` above.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+{
+    [m_state setObject:m_uniqueIdentifier forKey:lastSeenUniqueIdentifierKey];
+    writeStateToStorage();
+}
```
Can be removed. See other comment.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+NSNumber *WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage()
+{
+    return [currentState() objectForKey:lastSeenVersionStateKey];
```
`[m_state objectForKey:lastSeenVersionStateKey];` to avoid the copy.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    writeStateToStorage();
+}
+
+NSNumber *WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage()
```
Instead of returning an `NSNumber` return a C type like `NSInteger`. It lets the client avoid the unboxing.

xeenon replied:
Should still do this. If you hit an error with `NSInteger` try `uint64_t`.

xeenon reviewed (changes requested):
Looking good! Lots of small things.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -57,9 +57,13 @@
 // This number was chosen arbitrarily based on testing with some popular extensions.
 static constexpr size_t maximumCachedPermissionResults = 256;
 
+static constexpr NSString *backgroundPageListenersKey = @"BackgroundPageListeners";
```
Nit: `backgroundPageListenersStateKey`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1205,51 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        return;
+
+    if (!m_state.get().count)
+        return;
```
This should be removed. It won't save anything if state is empty, and we want to be able to save state the first time.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1136,8 +1142,21 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
-    loadBackgroundWebView();
+    if (!isPersistent()) {
+        NSNumber *backgroundPageListenersVersionNumber = loadBackgroundPageListenersVersionNumberFromStorage();
+        BOOL savedVersionNumberDoesNotMatchCurrentVersionNumber = backgroundPageListenersVersionNumber.integerValue != currentBackgroundPageListenerStateVersion;
```
Nit: Should use `bool`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1136,8 +1142,21 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        BOOL savedVersionNumberDoesNotMatchCurrentVersionNumber = backgroundPageListenersVersionNumber.integerValue != currentBackgroundPageListenerStateVersion;
+
+        // FIXME: <https://webkit.org/b/248889> Check to see if background page listens to onStartup().
+        BOOL backgroundPageListensToOnStartup = NO;
```
Nit: `bool` and `false`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1186,9 +1205,51 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+void WebExtensionContext::saveBackgroundPageListenersToStorage()
+{
+    if (backgroundContentURL().isEmpty() || isPersistent())
```
Use `extension().backgroundContentIsPersistent()`. `isPersistent()` is the wrong check, that is about the extension state and storage. No need to check `backgroundContentURL()` since `backgroundContentIsPersistent` does that internally.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
```diff
@@ -40,12 +41,14 @@
 void WebExtensionContext::addListener(WebPageProxyIdentifier identifier, WebExtensionEventListenerType type)
 {
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=248684.
+    if (!isPersistent() && m_backgroundWebView.get()._page->identifier() == identifier)
```
`!extension().backgroundContentIsPersistent()`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
```diff
@@ -40,12 +41,14 @@
 void WebExtensionContext::removeListener(WebPageProxyIdentifier identifier, WebExtensionEventListenerType type)
 {
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=248684
+    if (!isPersistent() && m_backgroundWebView.get()._page->identifier() == identifier)
```
`!extension().backgroundContentIsPersistent()`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1136,8 +1142,21 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 
-    // FIXME: <https://webkit.org/b/246483> Handle non-persistent background pages differently here.
-    loadBackgroundWebView();
+    if (!isPersistent()) {
```
`!extension().backgroundContentIsPersistent()`

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h:
```diff
@@ -63,7 +63,7 @@ WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))
 */
 + (instancetype)configurationWithIdentifier:(NSUUID *)identifier;
 
-@property (nonatomic, readonly, getter=isPersistent) BOOL persistent WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
I think we should keep the public API as-is. Just clarify the internal method. We will also need to be careful on changing the APIs after we go to API review. Otherwise we need to go to API review again.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
```diff
@@ -127,9 +127,9 @@ - (NSUUID *)identifier
+- (BOOL)isPersistentToStorage
 {
-    return _webExtensionControllerConfiguration->isPersistent();
+    return _webExtensionControllerConfiguration->isPersistentToStorage();
```
So just keep this part of the change.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -129,7 +130,7 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     NSError *createError(Error, NSString *customLocalizedDescription = nil, NSError *underlyingError = nil);
 
-    bool isPersistent() const { return hasCustomUniqueIdentifier(); }
+    bool isPersistentToStorage() const { return hasCustomUniqueIdentifier(); }
```
I know I threw this wording out yesterday, but I think `storageIsPersistent()` might be better, and better mirror the `backgroundContentIsPersistent()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
```diff
@@ -52,7 +52,7 @@ class WebExtensionControllerConfiguration : public API::ObjectImpl<API::Object::
     std::optional<UUID> identifier() const { return m_identifier; }
 
-    bool isPersistent() const { return !m_storageDirectory.isEmpty(); }
+    bool isPersistentToStorage() const { return !m_storageDirectory.isEmpty(); }
```
Ditto on `storageIsPersistent()`.

kiaraarose committed:
Add support for the Web Extension's Event API in the UIProcess.
https://bugs.webkit.org/show_bug.cgi?id=248684.

Reviewed by Timothy Hatcher.

This patch adds support for the Event API in the UIProcess. This includes loading the background
page listeners from storage when the extension loads, and saving these listeners to storage
(for non-persistent background pages).

* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm:
(-[_WKWebExtensionControllerConfiguration storageIsPersistent]):
Rename isPersistent() to storageIsPersistent() to make it clear that this function is for checking
to see if information relating to the extension is saved to storage and not for the extension
background page being persistent.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
(WebKit::WebExtensionContext::addListener):
(WebKit::WebExtensionContext::removeListener):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
Add FIXME to check to see if extension is being loaded as part of startup.

(WebKit::WebExtensionContext::stateFilePath const):
(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::writeStateToStorage const):
Use newly renamed storageIsPersistent().

(WebKit::WebExtensionContext::loadBackgroundWebViewDuringLoad):
Load background page listeners from storage.
Add FIXME to check to see if the extension is being loaded as part of startup.

(WebKit::WebExtensionContext::loadBackgroundPageListenersFromStorage):
Retrieve background page listeners data from storage. Populate m_backgroundPageListeners.

(WebKit::WebExtensionContext::saveBackgroundPageListenersToStorage):
Save background page listeners to storage for non persistent background pages.

(WebKit::WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage):
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads):
Save background page listeners to storage.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::storageDirectory const):

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::storageIsPersistent const):
(WebKit::WebExtensionContext::isPersistent const): Deleted.

* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::storageIsPersistent const):
(WebKit::WebExtensionController::isPersistent const): Deleted.

* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp:
(WebKit::WebExtensionControllerConfiguration::copy const):
Use newly renamed storageIsPersistent().

* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h:
(WebKit::WebExtensionControllerConfiguration::storageIsPersistent const):
(WebKit::WebExtensionControllerConfiguration::isPersistent const): Deleted.

Canonical link: https://commits.webkit.org/257589@main
